File: pracJava/파괴되지 않은 건물.java
class Solution {
    public int solution(int[][] board, int[][] skill) {
        int answer = 0;
        int m = board.length;
        int n = board[0].length;

        int[][] acc = new int[m+1][n+1];

        for(var s : skill){
            int degree = s[5];
            if(s[0] == 1){
                degree *= -1;
            }
            int y1 = s[1];
            int x1 = s[2];
            int y2 = s[3];
            int x2 = s[4];

            acc[y1][x1] += degree;
            acc[y1][x2+1] -= degree;
            acc[y2+1][x1] -= degree;
            acc[y2+1][x2+1] += degree;
        }

        for(int j = 0; j < m; j++){
            for(int i =1; i < n; i++){
                acc[j][i] += acc[j][i-1];
            }
        }


        for(int i = 0; i <n; i++){
            for(int j =1; j < m; j++){
                acc[j][i] += acc[j-1][i];
            }
        }

        for(int j = 0; j < m; j++){
            for(int i = 0; i < n; i++){
                board[j][i] += acc[j][i];
                if(board[j][i] > 0){
                    answer +=1;
                }
            }
        }


        return answer;
    }

    
}

File: baekjoonPython/프로그래머스/파괴되지 않은 건물.py
from collections import deque
def solution(board, skill):
    answer = 0
    
    # type 1 공격 / type 2 회복
    dx = [1,0,-1,0]
    dy = [0,1,0,-1]
    
    def BFS(ty, x1,y1, x2,y2, degree):
        
        isVisited = [[0]*len(board[0]) for _ in range(len(board))]
        
        q = deque()
        q.append([x1,y1])
        if ty == 1:
            degree *= -1
        while q:
            x,y = q.popleft()
            for i in range(4):
                xa = dx[i] + x
                ya = dy[i] + y
                
                
                if xa < 0 or ya < 0 or xa>=len(board) or ya>= len(board[0]) : continue
                if xa >x2 or xa < x1 or ya>y2 or ya<y1 or isVisited[xa][ya]: continue
                
                isVisited[xa][ya] = 1
                board[xa][ya] += degree
                q.append([xa,ya])
                
    
    for type, r1, c1, r2, c2, degree in skill:
        BFS(type, r1,c1, r2,c2,degree)
    
    
    for i in board:
        for j in i:
            if j > 0:
                answer +=1
    
    
    return answer